src/Type/ArrayType.php

```diff
@@ -354,7 +354,7 @@ public static function castToArrayKeyType(Type $offsetType): Type
 				return new IntegerType();
 			}
 
-			if ($offsetType instanceof StringType || $offsetType->isNonEmptyString()->yes()) {
+			if ($offsetType instanceof StringType || $offsetType->isNonEmptyString()->yes() || $offsetType->isLiteralString()->yes()) {
```

Comment:
Maybe the whole condition could be replaced with `$offsetType->isString()->yes()`?

Reply:
I tried so at first but that broke another test. I'll try to figure something out, because that is really how you'd want to write this.